### ISSUE TITLE
LaTeX: No spell check for chapter/section titles

### ISSUE THREAD

**Issue by ThelloD** [Jason-Rev/vscode-spell-checker#179]
The default language settings for LaTeX (located in "cspell-default.json") are a bit too harsh.

```
"languageSettings": [
    {
        "languageId": "latex",
        "dictionaries": ["latex"],
        "ignoreRegExpList": [
            "\\\\\\w*(\\[.*?\\])?(\\{.*?\\})?",
            "\\$.+?\\$"
            ]
    },
    ...
]
```

This ignores (custom) latex commands like `\doSomethng{arg}`, which is fine (no spell check needed for "doSometh(i)ng").

However, this rule also prevents spell check of titles of chapters and sections, for example:
`\subsection{Misspelld Title is not detectd!}`

My current solution is to shorten the first ignoreRegExp line:
```
"ignoreRegExpList": [
    "\\\\\\w*(\\[.*?\\])?\\{",
    "\\$.+?\\$"
]
```

However, this has the downside that code like `See section~\ref{implementationdescription}` will trigger a misspell warning ("implementationdescription" unknown).

Maybe there is a better solution?
My idea was to explicitly define include patterns for `\chapter`, `\section`, `\subsection`, but this didn't work. I suppose that the ignore patterns overwrite the include patterns.

Furthermore, I wasn't able to remove or overwrite these ignore patterns by altering my workspace settings. Plus, when I open the settings in vscode, on the left side (default settings), it tells me that no language setting is set, which is not true:
`"cSpell.languageSettings": [],`
In my opinion it would be better if the real default settings (stated above) are shown.

**Comment by Jason3S** [comment]
Hello Theo,

Thank you for taking the time to give feedback.

I too have been pondering how to solve that problem. It would be great to make sure this works: `\title{My greate woork of aart}`. I am looking into ways to solve this. But do not currently have a solution.

One of the things I am working on is the ability to nest include / excludes. At the moment that is not possible. Then doing what you suggest would work: *exclude all commands except ...*

Point well made on the default settings. As far as I know, VSCode does not allow for dynamic default settings.

[cspell](https://github.com/Jason3S/cspell) is the spell checker used by the extension. It is possible to test things out using the `cspell` command line. `cspell check myfile.tex` will show you which sections are excluded.

I just took a look at the LaTeX exclude definitions. I need to change them so you can override them. At the moment, they are hard wired. The fix is to use named patterns. It is possible to redefine a pattern in your own settings.

**Comment by James-Yu** [comment]
Hmm I'm thinking of refining regex for latex but cannot find the said `cspell-default.json` repo. Any guide @Jason3S ?

**Comment by Jason3S** [comment]
Hello @James-Yu,

Thank you. I have moved the LaTeX dictionary to [cspell-dicts](https://github.com/Jason3S/cspell-dicts). I have updated the regex patterns to include `chapter`, `section` etc. Please take a look.

I created named patterns, because they can be overridden if another pattern of the same name is created. This allows other people to fix any issues I may have created.

To test how well a pattern works, use [cspell](https://github.com/Jason3S/cspell) command line: `cspell check [-c <path to config file>] file_to_check.tex`

**Comment by Jason3S** [comment]
I hope to have an update to the spell checker this weekend.

**Comment by Jason3S** [comment]
This should be fixed. Please let me know.

**Comment by vallentin** [comment]
@Jason3S currently using 1.7.20, and this seems to not be working? Unless I'm doing something wrong.

Somewhat related, but is it possible to apply spell checking to custom commands? (In reference to #305.)

**Screenshot:**

![section](https://user-images.githubusercontent.com/17464404/69803044-4a488c80-11db-11ea-9ad0-84a423a79615.png)

**Comment by Jason3S** [comment]
@vallentin your function has the word `text` in the name. By default it will check command with the following words:
`title|color|section|subsection|footnote|chapter|part|caption|emph|text`

**Comment by vallentin** [comment]
@Jason3S wait, but why isn't `\section` and `\chapter` checked in my screenshot then? Like the `\text*` ones, I'd assume it would flag the misspelling.
